src/main/java/org/veupathdb/service/eda/us/model/DerivedVariableRow.java

```diff
@@ -0,0 +1,132 @@
+package org.veupathdb.service.eda.us.model;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import org.veupathdb.service.eda.generated.model.DerivedVariableGetResponse;
+import org.veupathdb.service.eda.generated.model.DerivedVariableGetResponseImpl;
+import org.veupathdb.service.eda.generated.model.DerivedVariablePostRequest;
+import org.veupathdb.service.eda.us.Utils;
+
+import java.time.OffsetDateTime;
+import java.util.Objects;
+
+import static org.gusdb.fgputil.functional.Functions.also;
+import static org.veupathdb.service.eda.us.Utils.mapIfPresent;
+
+/**
+ * Represents a complete database row for a Derived Variable.
+ */
+public class DerivedVariableRow {
+
+  public static final int MAX_DISPLAY_NAME_LENGTH = 256;
+  public static final int MAX_DESCRIPTION_LENGTH = 4000;
+
+  private final String variableID;
+  private final long userID;
+  private final String datasetID;
+  private final String entityID;
+  private final String displayName;
+  private final String description;
+  private final DerivedVariableProvenance provenance;
+  private final String functionName;
+  private final JsonNode config;
+
+  public DerivedVariableRow(
+    String variableID,
+    long userID,
+    String datasetID,
+    String entityID,
+    String displayName,
+    String description,
+    DerivedVariableProvenance provenance,
+    String functionName,
+    JsonNode config
+  ) {
+    this.variableID   = Objects.requireNonNull(variableID);
+    this.userID       = userID;
+    this.datasetID    = Objects.requireNonNull(datasetID);
+    this.entityID     = Objects.requireNonNull(entityID);
+    this.displayName  = Objects.requireNonNull(displayName);
+    this.description  = description;
+    this.provenance   = provenance;
+    this.functionName = Objects.requireNonNull(functionName);
+    this.config       = Objects.requireNonNull(config);
+  }
+
+  public DerivedVariableRow(String variableID, long userID, DerivedVariablePostRequest request) {
+    this(
+      variableID,
+      userID,
+      request.getDatasetId(),
+      request.getEntityId(),
+      request.getDisplayName(),
+      request.getDescription(),
+      null,
+      request.getFunctionName(),
+      Utils.JSON.convertValue(Objects.requireNonNull(request.getConfig()), JsonNode.class)
+    );
+  }
+
+  public DerivedVariableRow(String variableID, long userID, DerivedVariableRow copyFrom) {
+    this(
+      variableID,
+      userID,
+      copyFrom.datasetID,
+      copyFrom.entityID,
+      copyFrom.displayName,
+      copyFrom.description,
+      copyFrom.provenance != null ? copyFrom.provenance : new DerivedVariableProvenance(OffsetDateTime.now(), copyFrom.variableID),
```

Comment:
Sorry- one more comment; I think we should still use now() as the date on this new provenance since that's when it was copied.  So like:
```
new DerivedVariableProvenance(OffsetDateTime.now(),
   copyFrom.provenance != null
      ? copyFrom.provenance.getCopiedFrom()
      : copyFrom.variableID)
```

Reply:
I'm not sure I follow.  Are you saying replace line 77 with your suggestion?